modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ResizeHelper.java

```diff
@@ -67,81 +69,191 @@ public ResizeHelper(ResizeFeaturesBase rf,
 
         for (int i = 0; i < count; i++) {
             TableColumnBase<?,?> c = columns.get(i);
-            size[i] = c.getWidth();
+            size[i] = snapCeil(c.getWidth());
 
             if (c.isResizable()) {
-                double cmin = c.getMinWidth();
-                double cmax = c.getMaxWidth();
+                double cmin = snapCeil(c.getMinWidth()); // always honor min width!
+                double cmax = snapFloor(c.getMaxWidth()); // TableColumnBase.doSetWidth() clamps to (min,max) range
                 min[i] = cmin;
                 max[i] = cmax;
-                pref[i] = clip(c.getPrefWidth(), cmin, cmax);
+                pref[i] = clip(snapRound(c.getPrefWidth()), cmin, cmax);
+                // skip fixed columns
+                if (cmin == cmax) {
+                    skip.set(i, true);
+                }
             } else {
                 skip.set(i, true);
             }
         }
     }
 
     public void resizeToContentWidth() {
-        boolean needsAnotherPass;
-        do {
-            needsAnotherPass = false;
-            double sumWidths = 0.0;
-            double sumMins = 0;
-            for (int i = 0; i < count; i++) {
-                sumWidths += size[i];
-                sumMins += min[i];
+        if (skip.cardinality() == count) {
+            return;
+        }
+
+        // compute delta (snapped)
+        // if delta < 0 (and sum(width) < sum(pref)) -> distribute from size to min
+        // if delta < 0 (and sum(width) > sum(pref)) -> distribute from size to pref
+        // if delta > 0 (and sum(width) < sum(pref)) -> distribute from size to pref
+        // else -> distribute from size to max
+        //
+        double sumWidths = sum(size);
+        double sumPrefs = sum(pref);
+
+        double delta = target - sumWidths;
+        if (delta < 0.0) {
+            // shrink
+            if (target < sumPrefs) {
+                distribute(delta, min);
+            } else {
+                distribute(delta, pref);
+            }
+        } else if (delta > 0.0) {
+            // grow
+            if (target > sumPrefs) {
+                distribute(delta, max);
+            } else {
+                distribute(delta, pref);
             }
+        }
+    }
+
+    /** distibuting delta (positive when growing and negative when shrinking) */
+    private void distribute(double delta, double[] desired) {
+        double threshold = snapRound(SMALL_DELTA);
+        if (Math.abs(delta) > threshold) {
```

Comment:
The threshold is pretty arbitrary, I don't think it benefits from snap rounding here.  I'd be more inclined to eliminate the difference between these two functions; I would expect that if a user drags a column resizer for 100 pixels, that it should make no difference whether that was a slow drag or a fast one, and the end visuals and column sizes should be exactly the same regardless.

Reply:
theoretically, may be.  I could not achieve it with the algorithm that distributes space for large deltas, as it breaks down with small deltas, see my previous comment.